devito/arguments.py

```diff
@@ -158,198 +107,350 @@ class PtrArgument(Argument):
-        verify = verify and all([a.verify(v, enforce=enforce) for a, v in
-                                 zip(self.rtargs, value)])
+        for i in values:
+            verify = verify and all(verify(i, x) for x in i.verified_by)
         return verify
 
+    @property
+    def dimensions(self):
+        return [x for x in self.parameters if isinstance(x, Dimension)]
 
-class ConstantArgProvider(ArgumentProvider):
 
-    """ Class used to decorate Constant Data objects with behaviour required for runtime
-        arguments.
+class ArgumentVisitor(GenericVisitor):
+    """ Visits types to return their runtime arguments
     """
-
-    @cached_property
-    def rtargs(self):
-        return (ScalarArgument(self.name, self, default_value=self.data),)
-
-
-class TensorFunctionArgProvider(ArgumentProvider):
-
-    """ Class used to decorate Symbolic Data objects with behaviour required for runtime
-        arguments.
+    def visit_SymbolicFunction(self, o):
+        return TensorArgument(o)
+    
+    def visit_DimensionParameter(self, o):
+        dependency = Dependency("gets_value_from", o)
+        size = ScalarArgument(o.provider.size_name, [dependency])
+        start = ScalarArgument(o.provider.start_name, [dependency])
+        end = ScalarArgument(o.provider.end_name, [dependency])
+        return [size, start, end]
+    
+    def visit_Object(self, o):
+        return PtrArgument(o)
+
+    def visit_Array(self, o):
+        return TensorArgument(o)
+
+    def visit_Scalar(self, o):
+        dependency = Dependency("gets_value_from", o)
+        return ScalarArgument(o.name, o, dtype=o.dtype)
+
+    def visit_Constant(self, o):
+        # TODO: Add option for delayed query of default value
+        dependency = Dependency("gets_value_from", o)
+        return ScalarArgument(o.name, [dependency], dtype=o.dtype)
+
+    
+class ValueVisitor(GenericVisitor):
+    """Visits types to derive their value
     """
-
-    @cached_property
-    def rtargs(self):
-        return (TensorArgument(self.name, self),)
-
-
-class ScalarArgProvider(ArgumentProvider):
-
-    """ Class used to decorate Scalar Function objects with behaviour required for runtime
-        arguments.
+    def __init__(self, consumer, known_values):
+        self.consumer = consumer
+        self.known_values = known_values
+        super(ValueVisitor, self).__init__()
+        
+    def visit_Function(self, o, param=None):
+        assert(isinstance(self.consumer, TensorArgument))
+        return o.data
+
+    def visit_Constant(self, o, param=None):
+        return o.data
+
+    def visit_Dependency(self, o):
+        return self.visit(o.obj, o.param)
+
+    def visit_function(self, o, param=None):
+        return o(self.consumer, self.known_values, param)
+
+    def visit_Object(self, o, param=None):
+        return o.value
+
+    def visit_object(self, o, param=None):
+        return o
+
+    def visit_DimensionParameter(self, o, param=None):
```

Comment:
this is a mess to understand...

Reply:
I know. I'm trying to figure out a way to simplify this.